### PR TITLE
Deprecate SequentialGreedyRecommender

### PR DESCRIPTION
**Pull request by AdrianSosic**
This PR deprecates the `SequentialGreedyRecommender` class and replaces it with the new `BotorchRecommender` class, which now let's the user control the `sequential` flag being passed to the underlying optimization routine.

### CODE REVIEW

File: baybe/recommenders/pure/bayesian/botorch.py
Comment (Scienfitz) [inline baybe/recommenders/pure/bayesian/botorch.py:1]: because of the file move its hard to see the differences, I see the `sequential` flag addition (expected), is there nay other thing that was changed worth mentioning?

File: baybe/recommenders/pure/bayesian/sequential_greedy.py
Comment (Scienfitz) [inline baybe/recommenders/pure/bayesian/sequential_greedy.py:17]: technically this should mention that the sequential flag must be True for a 1to1 replacement

File: docs/userguide/recommenders.md
Comment (Scienfitz) [inline docs/userguide/recommenders.md:41]: the new `sequential` flag together with a sentence of the difference between true and false needs to be mentioned here (possibly even more important than the sampler stuff

**Review by Scienfitz — COMMENTED** [review]
good to have this glattgezogen!

File: baybe/recommenders/pure/bayesian/sequential_greedy.py
Comment (AdrianSosic) [reply baybe/recommenders/pure/bayesian/sequential_greedy.py:17]: Technically: no, in fact :D Because the previous behavior WAS the one with `sequential=True` (i.e. bug). That is, you'd really get the old behavior when doing just nothing and using the `sequential=False` default. Makes sense?

File: baybe/recommenders/pure/bayesian/botorch.py
Comment (AdrianSosic) [reply baybe/recommenders/pure/bayesian/botorch.py:1]: Not really, apart from a slight reformatting of the docstring. But note: I did the movement in a single commit (on purpose) so that you can very easily see the diff if you compare not against the root but against the first commit of this PR. I always do it like that so that people have an easier time reviewing. Let me know if you have more questions or close 🙃

File: baybe/recommenders/pure/bayesian/base.py
Comment (AdrianSosic) [reply baybe/recommenders/pure/bayesian/base.py:31]: Not "necessary" but I noticed that the signature was not very nice:
image
The logical order to me is (ignoring the deprecated argument): 1) Surrogate 2) ACQF 3) Rest

And this is indeed what you get after the change:
image

We can discuss though whether we want `sequential` as `kwonly` or not. Right now, I've made it keyword. Preferences?

File: baybe/recommenders/pure/bayesian/sequential_greedy.py
Comment (Scienfitz) [reply baybe/recommenders/pure/bayesian/sequential_greedy.py:17]: to follow up: imo this is both
1) a refactoring, renaming the recommender name
2) a bug fix

Imo, we do not need to mix 1 and 2, the error should not carry over into the deprecation.

In fact the README should have an entry in Fixed that says seq greedy recommender sequential flag not being activated

File: baybe/recommenders/pure/bayesian/base.py
Comment (Scienfitz) [reply baybe/recommenders/pure/bayesian/base.py:31]: ok seems alright

on that not: why is `acquisition_function_cls` a `bool` and why is it not `Optional` ? I know that was not introuced in this PR and its the deprecated thing, but that still seems wrong?

Comment (AdrianSosic) [reply baybe/recommenders/pure/bayesian/base.py:31]: Well, you may ask that yourself :D Here the changes from the commit you made in the acqf refactoring:
image
image
So I guess you copy-pasted the logic from the campaign deprecation I introduced
image
but did not adjust the types, I and didn't bother during review since it's a deprecation anyway. But I'm happy to fix it. In fact, have just pushed a commit. Please confirm if my explanation was correct and resolve if happy

File: baybe/recommenders/pure/bayesian/sequential_greedy.py
Comment (AdrianSosic) [reply baybe/recommenders/pure/bayesian/sequential_greedy.py]: I know what you mean, and if you really prefer that, I will not block it (have pushed a commit that you can inspect and see if that's what you want). But the downsides I see about this is that the consequences could be difficult to disentangle for the user. With my previous approach, they wouldn't need to bother: upgrading to `BotorchRecommender` would really give them the same results as before, and everything becomes automatically bug-free. With your approach, on the other hand:
* They would see that they need to upgrade, but the deprecation message would explicitly tell them to active `sequential=True`
* This would NOT be a 1:1 replacement to what they had before, giving different results.
* Instead, they would also need to know (magically, unless they look into the changelog, by chance) that `sequential=True` is a bugfix and not actually what they need to do for backward-compatibility.

In fact, when I read the new deprecation warning myself, I was also a bit confused first 😄 

So, in a sense, you are asking them to break their own backward-compatibility, and the changelog mentions a bug that is no longer relevant to the users, since they would anyway upgrade to `BotorchRecommender` as the next step. With the previous approach, we'd simply skip that "inconsistent stage" altogether, keeping the old behavior but now with a class label that no longer makes it a bug.

Comment (Scienfitz) [reply baybe/recommenders/pure/bayesian/sequential_greedy.py]: the downside of your proposed way
- providing the user a `SequentialGreedyRecommender` that is not sequential

seems to outweigh all your points and I dont see why providing a deprecation 1to1 replacement for something should take precedence over providing `what the user has requested when they wrote the word Sequential`. Given that we have no evidence of any more computational expense or so, what do you mean with `user has to bother`?

I see that calling it a `1to1` replacement doesnt make sense though, just use `please use BotorchRecommender(sequential=True)` or similar

Comment (AdrianSosic) [reply baybe/recommenders/pure/bayesian/sequential_greedy.py]: As I said, I'm fine with your proposal because it's a deprecation anyway, so I "accept" :D But I in fact noticed a much more important issue – and I think that is the reason why we called it `sequential` back at times when we only supported discrete spaces: the `sequential` flag is relevant only for conti spaces. For discrete/mixed spaces, it automatically performs sequential optimization, whereas for conti spaces, sequential is `False` by default.

Given the above, I think it's confusing to just expose `sequential=False` as default, as it would be quite misleading for non-conti spaces. Would you be fine if I instead set it to `sequential: bool | None = None` and also raise the corresponding warnings at runtime (e.g. warning the user that his `sequential` flag being set to `False` will be ignored once he uses the recommender in a non-conti setting)? Will prepare the change and you can then simply review it

Comment (AdrianSosic) [reply baybe/recommenders/pure/bayesian/sequential_greedy.py]: Updated logic is now there. Please let me know what you think / resolve

File: docs/userguide/recommenders.md
Comment (AdrianSosic) [reply docs/userguide/recommenders.md:41]: Added a short paragraph. Let me know if something is missing.

File: baybe/recommenders/pure/bayesian/sequential_greedy.py
Comment (Scienfitz) [reply baybe/recommenders/pure/bayesian/sequential_greedy.py]: if `sequential` only applies to discrete spaces, why dont we make that clear in the flag name and make the logic even simpler (ie not so many warnings or explanations needed)?

Comment (AdrianSosic) [reply baybe/recommenders/pure/bayesian/sequential_greedy.py]: Also OK for me. Do you have a suggestion?

Comment (Scienfitz) [reply baybe/recommenders/pure/bayesian/sequential_greedy.py]: `sequential_discrete`

Comment (Scienfitz) [reply baybe/recommenders/pure/bayesian/sequential_greedy.py]: sorry I prob misread, if its only relevant in conti spaces it should be `sequential_continuous`, right?

Comment (AdrianSosic) [reply baybe/recommenders/pure/bayesian/sequential_greedy.py]: Yeah, noticed it, too :D Now I've fixed it

File: CHANGELOG.md
Comment (Scienfitz) [inline CHANGELOG.md:8]: forgotten something

File: docs/userguide/recommenders.md
Comment (Scienfitz) [inline docs/userguide/recommenders.md:38]:
```suggestion
  selection is the only available mode and is thus activated automatically. 
  
  Note that
```
(suggesting to split the paragraph)

**Review by Scienfitz — APPROVED** [review]

File: docs/userguide/recommenders.md
Comment (Scienfitz) [reply docs/userguide/recommenders.md]: gucci

File: CHANGELOG.md
Comment (AdrianSosic) [reply CHANGELOG.md]: thx, changed it (also added an additional split at the end, since the grouping would have been weird otherwise)

File: docs/userguide/recommenders.md
Comment (AdrianSosic) [reply docs/userguide/recommenders.md]: Thanks, also added a second split at the end, since the grouping would have been weird otherwise

File: CHANGELOG.md
Comment (AdrianSosic) [reply CHANGELOG.md]: Added "Added" :D

File: CHANGELOG.md
Comment (AdrianSosic) [inline CHANGELOG.md:16]: @Scienfitz: OK, PR ready to go 🚀  Last question: are we now happy with `BotorchRecommender`? I made my peace with it, because it is the most accurate name I saw until now, but also want your final approval 🙃

Comment (Scienfitz) [reply CHANGELOG.md:16]: sure

**Pull request merged**
